### PR TITLE
[Will Not Merge] Logins API Shape

### PR DESCRIPTION
**Pull request by lougeniaC64**
TL;DR This branch is a tool, it will not be merged.

I created this PR as a means to socialize [the updated Logins API](https://github.com/mozilla/application-services/blob/8c5950031b256ad44ad6e66d833dd114f83f6349/components/logins/src/logins.udl) and provide a stable branch for the work required for issues #4199 and #4200. There are a lot of dubious code changes here (not to mention commented out tests 😵) that were made solely to allow parallelizing the logins migration, API, and consumer code updates. I will make [the necessary logins API changes](#4091) in earnest in a forthcoming PR. Once I receive initial feedback on this API shape, I'll close this PR leaving the `logins-api-shape` branch available for the aforementioned consumer work.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/logins/src/logins.udl [hunk 1]
```diff
@@ -2,18 +2,26 @@
...
+    i64 times_used;
+    i64 time_created;
+    i64 time_last_used;
+    i64 time_password_changed;
 };
 
-
 // Converted from mozilla.appservices.logins.protobuf.rs
-dictionary Login {
+dictionary LoginPayload {
```
Comment (skhamis) [inline components/logins/src/logins.udl:24]: I like the idea of `Login` and `LoginPayload` given that we can have the payload be a reduced set of params that is needed to be sent over by the consumer, however I see that `LoginPayload` still contains all the fields -- will this be going away? I would imagine it would like something like:

```
LoginPayload {
    string id;
    string hostname;
    string password;
    string username;
    string? http_realm;
    string? form_submit_url;
    string username_field;
    string password_field;
}
```

but I'm probably missing something here, mainly my initial thoughts from my current mental map!

Comment (lougeniaC64) [reply components/logins/src/logins.udl:24]: If you look in the login.rs file of the per-field-encryption branch, [the Login struct](https://github.com/mozilla/application-services/blob/18e48dd16b07272a6519a4fe0844d3d3a98eb918/components/logins/src/login.rs#L241) has the encrypted username and password fields and the `LoginPayload` struct is for the unencrypted fields. I do like separating those out so the consumers understand which data they're dealing with. But also, I think that the `username_enc` and `password_enc` fields will end up being optional strings, or diverging from `Logins` in some other way (particularly when the scrubbing logic is added). Having said that, it may be that using one struct is fine for now or that `LoginPayload` isn't the correct nomenclature. Either way, open to suggestions

Comment (bendk) [reply components/logins/src/logins.udl:24]: `LoginPayload` was designed to be the data we got from the server, which was why the username/password were plaintext.  I think we need to keep that struct as-is (or maybe someone who knows serde more can implement that code in a batter way).

I do like idea of a `Login`-like struct with reduced fields that's intended for the consumer code.  It would just need to be separate from `LoginPayload`

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt [hunk 2]
```diff
@@ -111,60 +111,40 @@ class DatabaseLoginsStorage(private val dbPath: String) : AutoCloseable {
     }
 
     @Throws(LoginsStorageException::class)
-    fun list(): List<Login> {
+    fun decryptAndFixupLogin(encKey: String, login: Login): LoginPayload {
```
Comment (bendk) [inline components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt:114]: Maybe we can commit to this just being `decrypt()` since Sammy is going to rewrite the migration to validate it?  I'm not sure on that at all, just wanted to raise the question.

File: components/logins/src/logins.udl (hunk 1 above)
Comment (lougeniaC64) [reply components/logins/src/logins.udl:24]: I'm not sure which fields are expected and which we can hide, I know that `time_password_changed` needs to be exposed for iOS and I think one of the other timestamp fields might be used for ordering.

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt [hunk 3]
```diff
@@ -111,60 +111,40 @@ class DatabaseLoginsStorage(private val dbPath: String) : AutoCloseable {
...
-            checkUnlocked().add(login)
+    fun getByBaseDomain(baseDomain: String): List<Login> {
+        return readQueryCounters.measure {
+            checkUnlocked().getByBaseDomain(baseDomain)
         }
     }
 
     @Throws(LoginsStorageException::class)
-    fun importLogins(logins: List<Login>): String {
+    fun addOrUpdate(encKey: String, login: Login): String {
```
Comment (bendk) [inline components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt:135]: `addOrUpdate()` is the biggest question for me.  I'm going to copy/rephrase my thoughts from the google doc here.  BTW, thanks for making this PR, it's nice to have the design more in the open.

Based on the way the API is currently shaped, it seems like you should use the API like this:

 - When the user clicks on the username/password field, call getByBaseDomain() and present some of them to the user
 - If the user clicks on one, then autofil the form and store that `Login` somewhere
 - If the user changes the username/password field values, then update the username/password fields for that stored `Login`
 - If the user submits the form and clicks "Save this login", send the stored `Login` to `addOrUpdate()`.  `addOrUpdate()` will then update that `Login` since it already existed in the DB.

However, depending on what the user entered into the username field, we aren't going to update that login.  If they enter a new username then we should insert a new record.  If they enter an existing username, then we should update a different record. 

I think the issue is that `addOrUpdate()` seems like it would work like a `save()` method in an ORM: it will do an `INSERT` if the guid is not set and an `UPDATE ... WHERE guid={guid}` if the guid it set.  But actually, the GUID is more like a hint.  It only matters as a tiebreaker when there are duplicate records with the same username.   (At least this how I'm reading  [shouldUpdateOrCreateAsync()](https://github.com/mozilla-mobile/android-components/blob/d7b1b62da50cf224fce1eec7a163d858fb1a2334/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L29) and [ensureValidLogin()](https://github.com/mozilla-mobile/android-components/blob/d7b1b62da50cf224fce1eec7a163d858fb1a2334/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L74))

So I think the signature should change to better indicate what's happening:
  - Define a `LoginData` dictionary, which is the same as `Login`, but without the GUID field
  - Make the signature `addOrUpdate(encKey: String, login: LoginData, selectedGuid: String?)`

Also, on the consumer side we need to store the `Login` the user selects, but we shouldn't be mutating it.  Instead, we should mutate a separate `LoginData` value.  This is because we don't really know which `Login` is going to be updated until the entire process is complete.

**Comment by bendk** [comment]
This is great overall, it was really nice for me to have something more concrete to discuss.  I think the only thing that needs serious thought is addOrUpdate(), my other comments address fairly minor things.

File: components/logins/src/logins.udl (hunk 1 above)
Comment (lougeniaC64) [reply components/logins/src/logins.udl]: I made some updates to the exposed login structs. I removed `LoginPayload`, added a `LoginFields` struct with the meta fields removed for the `add_or_update` function, and then added a `DecryptedLogin` struct for the `decrypt_and_fixup_login` function. The meta fields are still needed for the fennec to fenix migration code, but we can consider removing those when we stop supporting that logic.

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt (hunk 2 above)
Comment (lougeniaC64) [reply components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt:114]: We should postpone this decision until after the current effort so shipping the migration isn't delayed and any fixup-related regressions are isolated. But I hope that we are able to eventually remove the fixup logic here.

File: components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt (hunk 3 above)
Comment (lougeniaC64) [reply components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt:135]: IIUC the login returned from ` shouldUpdateOrCreateAsync` is merged [here](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt#L76) with the selected login (the one passed into the [onLoginSave function](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt#L68) where `shouldUpdateOrCreateAsync` is called via `ensureValidAsync`) and the resulting merged login is then passed to the [update function](https://github.com/mozilla-mobile/android-components/blob/bfaf514793dd20f9f12d3f551f9ff3e470f52727/components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt#L73). So while I agree the login guid is used as more of a tool to assess the precedence of the potential dupes in the `shouldUpdateOrCreateAsync` function, the login guid dictates which login is updated thereafter making it more consequential. 

Outside of the potential delays the necessary product and/or external team dependencies required to make this update, I question whether this change should be included in an effort that should be invisible to the user. Having said that, I think it's worth capturing this in a ticket for future consideration.

Comment (bendk) [reply components/logins/android/src/main/java/mozilla/appservices/logins/DatabaseLoginsStorage.kt]: My thinking was that we were already going to have to change the consumer code, so it wouldn't be much extra effort.  But if that's not true then I definitely agree that we should try to limit the scope.  In that case, what about adding some comments to the `addOrUpdate()` that describe how it's used and what the GUID field is doing?

**Comment by lougeniaC64** [comment]
I think the discussion we're about to have makes this PR obsolete so I'm closing it. cc: @bendk @mhammond
